### ISSUE TITLE
Question: What's the difference with electron-forge?

### ISSUE THREAD

**Issue by mrpatiwi** [electron-userland/electron-builder#1193]
I'm building a Electron app and I just found this Github organization with two interesting projects:

* electron-builder
* electron-forge

But I can't find information about their difference 😞 

* What are the _pros_ and _cons?
* How do they relate to https://github.com/electron/electron-compile?
* Does support Babel out-of-the-box?
* Does they support minify, Tree-shaking, Dead-code-elimination? How about Node.js native dependencies?
* Which one comes with less boilerplate?

Any help is appreciated.

**Comment by zeke** [comment]
cc @malept @MarshallOfSound

**Comment by develar** [comment]
electron-builder it is a tool to build your Electron application. 

electron-forge it is a tool to build your Electron application PLUS boilerplate to create electron app.

 > What are the pros and _cons?

electron-builder supports 
* code signing (CI support, auto discovery, multi-platform windows code sign)
* publishing (including Bintray, S3 support coming soon)
* easy way to configure files to pack ([files](https://github.com/electron-userland/electron-builder/wiki/Options#Config-files), extraFiles and extraResources). 
* Numerous targets (including AppImage, Snaps, pkg and so on).
* Build all targets using one unified interfaces.
* [server-less auto update](https://github.com/electron-userland/electron-builder/wiki/Auto-Update)

> How do they relate to https://github.com/electron/electron-compile?

electron-builder and electron-forge are tools to _build_ your electron app. electron-compile to _create_.

> Does support Babel out-of-the-box?

For electron-builder it is out of scope, for electron-forge as well (so, electron-forge uses it under the hood).

> Does they support minify, Tree-shaking, Dead-code-elimination? How about Node.js native dependencies?

Both tools supports native node deps management. electron-builder also supports Yarn. Maybe electron-builder does it better, but it depends. In any case you can use electron-rebuild (this tool is used under the hood by electron-forge).

> Which one comes with less boilerplate?

electron-builder. Because electron-forge also helps you to create your app, not only to build. But electron-forge also doesn't create tons of boilerplate code, as far I know.

**In short: you don't need to choose or only electron-forge, or only electron-builder.** electron-builder doesn't restrict you and you can use any framework and any packaging tool. You can use electron-builder to [pack only in a distributable format](https://github.com/electron-userland/electron-builder#pack-only-in-a-distributable-format).

**Comment by MarshallOfSound** [comment]
I'll answer the questions the same way as @develar 👍 

> What are the pros and _cons?

`electron-forge` supports

* code signing
* publishing (github, with bintray and S3 coming soon)
* maximum customizability, we just wire up other packages together, you can pass whatever config you want, to **any** of our dependencies.  You retain full control of your own build process
* Numerous targets as well 😆 Zip, DMG, Squirrel, Appx, Flatpack and so on
* Start, Build, Package, Lint and do everything through Electron Forge while still retaining all the customizability mentioned above.  Benefit of this is things like native modules work seamlessly in your development environment as well as when packaging your application
* Support auto update through the standard Electron Auto-Update compatible artifacts that Squirrel.Windows and Squirrel.Mac require.

> How do they relate to https://github.com/electron/electron-compile?

`electron-forge` not only builds your application it helps you create them as well, by using `-compile` under the hood you can use futuristic JS features with **zero** config on your part.  You can even use cool languages like TypeScript, LESS, SCSS and Jade templates with no config (they **just work**).  We also have built in templates for `react`, `react-typescript`, `angular2` and `vue` so it's super easy to get started with a new Electron application.

> Does support Babel out-of-the-box?

See above 👍 yes `electron-forge` supports babel transpilation (though `electron-compile` out of the box but it is done in a way so you don't even realize babel is at work, it is seamlessly done in the background for a much nicer development experience.

> Does they support minify, Tree-shaking, Dead-code-elimination? How about Node.js native dependencies?

Minification and Tree-Shaking are both out of scope of `electron-forge` currently but would not be too hard to add using the packager hooks provided by `electron-packager`.  As for Native Modules we support **all** native modules and we also built `electron-forge` from the ground up with the intention of using both `npm` and `yarn ` to develop Electron application so we also support `yarn` and `npm` throughout the entire development lifecycle.

> Which one comes with less boilerplate?

Technically speaking `electron-builder` comes with less boilerplate (it comes with no boilerplate) as it is intended to be solely for **building** electron applications.  Electron Forge on the other hand is intended to help you **create** them as well and therefore we provide **ultra minimal** boilerplating to help you get started.  On average this is like 6 files

* package.json (with prefilled forge config)
* compilerc (for electron-compile)
* index.html (template HTML file)
* index.js (template main process file)
* eslintrc (for linting)
* maybe another file (react has app.jsx for instance)

It is designed to be ultra minimal and give you minimum required files to get you going on your Electron way 👍 

**Bonus Features**

We also built `electron-forge` to be extensible in a real easy way.  For instance anyone can publish `electron-forge-template-[name]` modules and use them as templates when making a new Electron application.  Or people can publish `electron-forge-maker-[name]` to provide a new make target or even `electron-forge-publisher-[name]` to provide a new publish target.  We also also passthrough directly to the hooks of the tools we use like `electron-packager`'s lifecycle hooks.

**Summary:**

`electron-forge` was built to provide a wider scope than `electron-builder` ever intended to as it's goal is not just to build Electron application but to help you from the moment you create your app to the moment you want to publish it.  Our main goal was to make a tool that used community packages that we already used and wire them up in an intuitive and easy to use way so that **anyone** could just jump right in to Electron development without having to worry about setting up build files and such.

I believe we achieved this goal when our commands are as simple as

```bash
npm i -g electron-forge
electron-forge init my-app

cd my-app
electron-forge make
```

**Comment by develar** [comment]
electron-builder 16.2.0 supports electron-forge and electron-compile.

You can use appimage, snap, nsis and nsis-web electron-builder targets in the electron-forge. But it is not recommended, because electron-forge packaging and publishing is a far away in terms of quality / feature set from electron-builder.

Recommended to build electron-forge projects using electron-builder directly. [Publishing](https://github.com/electron-userland/electron-builder/wiki/Publishing-Artifacts), [Auto Update](https://github.com/electron-userland/electron-builder/wiki/Auto-Update) and [Code Signing](https://github.com/electron-userland/electron-builder/wiki/Code-Signing) supported only in this case (including all other useful [packaging options](https://github.com/electron-userland/electron-builder/wiki/Options) like [files](https://github.com/electron-userland/electron-builder/wiki/Options#Config-files)). 

As it was stated initially, electron-forge is not a competitor to electron-builder. If you are happy to use electron-forge to package/publish — your decision, in any case you still can use electron-builder targets. User requests to improve electron-builder integration with electron-forge (targets) will be not rejected as "just use electron-builder directly" without investigation.

I don't like situation that electron-forge decided to reinvent the wheel and users have to solve already solved and implemented issues (https://github.com/electron-userland/electron-forge/issues/171, https://github.com/electron-userland/electron-forge/issues/135), but it was already discussed. Up to user decide what to use to build. electron-forge maybe useful to create electron app and, so, electron-builder 16.2.0 supports electron-forge.

**Comment by trusktr** [comment]
The fact that you can use both for either-or certain things, may be confusing for someone new. The question is, what specifically would you use electron-forge for, and what specifically electron-build for (if using both).

Does anyone have intimate experience with both tools, and can give a comparison of (or how they complement) each other based on that experience?

Basically what a newb wants to know is "what is the best workflow for building Electron apps for all platforms (whether that's using one tool or both tools), in the opinion of someone that intimately knows how to use both tools?".

I suppose I will try both of them out, and come back to post if someone else hasn't by the time I do. :)

**Comment by tmtron** [comment]
@trusktr Did you already test both and have some news?

**Comment by cyraid** [comment]
> @trusktr Did you already test both and have some news?

I tried both. I'll give you my experiences.

electron-builder: Tough to get rolling, gotta start a project first, ran into some problems. Shares the dist folder, gotta add a 'files' to package.json, slow process, needed to clone a project for typescript.

electron-forge: Ran `npx create-electron-app my-new-app --template=typescript`, didn't need to do anything more, did a start, boom. Worked. Did a 'package', and quite quickly, it made a directory (in out, not dist), and everything just.. worked.

So, based on my initial workings, electron-forge is by far the better one (good job Forge team btw).

**Comment by fjeddy** [comment]
So your basing what's the better solution on your own lack of knowledge and ignorance @cyraid ? It's incredible how todays "developers" determine what's good or bad solely on how little they have to do to do something.

The internet is getting more and more bloated by applications and websites that run on frameworks that run on frameworks that run on frameworks on top of another framework, nothing is optimized for the end user, speed and stability, no, everything is optimized so @cyraid has to do as little work as possible.

**Comment by cyraid** [comment]
> So your basing what's the better solution on your own lack of knowledge and ignorance @cyraid ? It's incredible how todays "developers" determine what's good or bad solely on how little they have to do to do something.
> 
> The internet is getting more and more bloated by applications and websites that run on frameworks that run on frameworks that run on frameworks on top of another framework, nothing is optimized for the end user, speed and stability, no, everything is optimized so @cyraid has to do as little work as possible.

Right, because build tools are indeed meant to be in the way as little as possible. Do you wanna start up a new project only to find you have to mess around with the compiler/build tools for hours first? Or would you like to start your new project right away while you still have what's in your mind to put into it?

Like I said, I tried both and gave my experiences. My insight was that of a new User of the Framework. Sharing insight on which took more work to get started.

**Comment by FunctionDJ** [comment]
> So your basing what's the better solution on your own lack of knowledge and ignorance @cyraid ? It's incredible how todays "developers" determine what's good or bad solely on how little they have to do to do something.
> 
> The internet is getting more and more bloated by applications and websites that run on frameworks that run on frameworks that run on frameworks on top of another framework, nothing is optimized for the end user, speed and stability, no, everything is optimized so @cyraid has to do as little work as possible.

Maybe you're right, but your neither polite nor helpful.

**Comment by akash07k** [comment]
True, not polite at all.
> > So your basing what's the better solution on your own lack of knowledge and ignorance @cyraid ? It's incredible how todays "developers" determine what's good or bad solely on how little they have to do to do something.
> > The internet is getting more and more bloated by applications and websites that run on frameworks that run on frameworks that run on frameworks on top of another framework, nothing is optimized for the end user, speed and stability, no, everything is optimized so @cyraid has to do as little work as possible.
> 
> Maybe you're right, but your neither polite nor helpful.
